### PR TITLE
fix bug: capability validation failed when some running jobs finished or deleted

### PR DESCRIPTION
**Pull request by Thor-wl**
bug fix point:
1.  the right time of capability validation is in enqueue action(required validation), not the jobEqueueableFn of plugins
2.  allocated resource for queue calculated incorrectly. Considering this cenario that some jobs are in the same queue, while parts are running but parts are pending because of capability limit. When one of running job finished or deleted, next scheduler cycle will choose some pending jobs (actually podgroups) inqueue by go through all pending jobs in the queue. In each inspect period for one job, scheduler will attempt to compare the sum of allocated reource and this job request resource with capability. If the inqueue jobs request resource are not added to allocated, every check round will get a same allocate resource. In fact, ahead inqueue jobs has already locked in some resource quota in capability. 

Signed-off-by: Thor <1187526662@qq.com>

### CODE REVIEW

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 1]
```diff
@@ -142,3 +143,47 @@ func (enqueue *Action) getOverCommitFactor(ssn *framework.Session) float64 {
...
+							allocatedResource.Add(t.Resreq)
+						}
+					}
+				}
+			}
+		}
+	}
+	minReq := api.NewResource(*targetJob.PodGroup.Spec.MinResources)
+	klog.V(3).Infof("minResources of podgroup of job %s: %s(cpu)", targetJob.Name, strconv.FormatFloat(minReq.MilliCPU, 'E', -1, 64))
+	klog.V(3).Infof("podgroup allocated: %s(cpu)", strconv.FormatFloat(allocatedResource.MilliCPU, 'E', -1, 64))
```
Comment (hzxuzhonghu) [inline pkg/scheduler/actions/enqueue/enqueue.go:187]: ditto

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 2]
```diff
@@ -142,3 +143,47 @@ func (enqueue *Action) getOverCommitFactor(ssn *framework.Session) float64 {
...
+	// If no capability is set, always enqueue the job.
+	if len(queue.Queue.Spec.Capability) == 0 {
+		klog.V(4).Infof("Capability of queue <%s> was not set, allow job <%s/%s> to Enqueue.",
+			queue.Name, targetJob.Namespace, targetJob.Name)
+		return true
+	}
+
+	// If capability is set, calculate allocated resource of queue which the target belongs to,
+	// compare the sum of allocated and Job minRequest resource with the capability resource
+	allocatedResource := api.EmptyResource()
```
Comment (hzxuzhonghu) [inline pkg/scheduler/actions/enqueue/enqueue.go:160]: I think this way is very inefficient as for every job, it is calculated once
suggest adding a new field in `queueAttr`

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 3]
```diff
@@ -142,3 +143,47 @@ func (enqueue *Action) getOverCommitFactor(ssn *framework.Session) float64 {
 
 	return factor
 }
+
+func isCapabilitySatisfied(ssn *framework.Session, targetJob *api.JobInfo) bool {
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go:147]: it's better to handle in callback function.

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 4]
```diff
@@ -110,12 +109,14 @@ func (enqueue *Action) Execute(ssn *framework.Session) {
...
-
 		if job.PodGroup.Spec.MinResources == nil {
 			inqueue = true
 		} else {
 			minReq := api.NewResource(*job.PodGroup.Spec.MinResources)
-			if ssn.JobEnqueueable(job) && minReq.LessEqual(idle) {
+			klog.V(3).Infof("minResources of podgroup of job %s: %s(cpu)", job.Name, strconv.FormatFloat(minReq.MilliCPU, 'E', -1, 64))
+			klog.V(3).Infof("node idle: %s(cpu)", strconv.FormatFloat(idle.MilliCPU, 'E', -1, 64))
+			klog.V(3).Infof("JobEnqueueable: %t", ssn.JobEnqueueable(job))
+			if isCapabilitySatisfied(ssn, job) && ssn.JobEnqueueable(job) && minReq.LessEqual(idle) {
```
Comment (k82cn) [inline pkg/scheduler/actions/enqueue/enqueue.go:119]: how about the performance?

File: pkg/scheduler/actions/enqueue/enqueue.go [hunk 5]
```diff
@@ -81,15 +81,14 @@ func (enqueue *Action) Execute(ssn *framework.Session) {
...
-
 	total := api.EmptyResource()
 	used := api.EmptyResource()
 	for _, node := range ssn.Nodes {
 		total.Add(node.Allocatable)
 		used.Add(node.Used)
 	}
 	idle := total.Clone().Multi(enqueue.getOverCommitFactor(ssn)).Sub(used)
-
+	klog.V(3).Infof("used cpu: %s, idle cpu: %s", strconv.FormatFloat(used.MilliCPU, 'E', -1, 64), strconv.FormatFloat(idle.MilliCPU, 'E', -1, 64))
```
[the remark this answers is not in the document]
Comment (Thor-wl) [reply pkg/scheduler/actions/enqueue/enqueue.go]: done

File: pkg/scheduler/actions/enqueue/enqueue.go (hunk 2 above)
Comment (Thor-wl) [reply pkg/scheduler/actions/enqueue/enqueue.go:155]: Maybe two filed "inqueueSource" & "allocatedSource" is a good way, i need to implement it after sorting out of state transition and resource computational logic.

**Comment by Thor-wl** [comment]
create a new issue to fix in a new way at issue 959

**Comment by Thor-wl** [comment]
create a new issue to fix in a new way at issue959

**Pull request closed without merging**